File: dashboard/backend/app/schemas/matching.py
"""
Pydantic schemas for matching API endpoints.
Supports 1:N relationship model (NAS File → Sheet Records → UDM Segments).
"""
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field


class NasFileInfo(BaseModel):
    """NAS file information (Asset Level)"""
    exists: bool
    path: str
    size_mb: float = Field(description="File size in megabytes")
    duration_sec: float | None = Field(None, description="Video duration in seconds")
    modified_at: datetime
    inferred_brand: str | None = None


class UdmInfo(BaseModel):
    """UDM segment conversion info"""
    uuid: str | None = None
    status: Literal["complete", "pending", "warning", "error"] = "pending"


class SegmentRecord(BaseModel):
    """
    Individual segment record (Segment Level).
    One NAS file can have multiple segments (1:N relationship).
    """
    row_number: int = Field(description="Sheet row number")
    source: Literal["archive_metadata", "iconik_metadata"]

    # Time information
    time_in: str = Field(description="Timecode (HH:MM:SS)")
    time_out: str = Field(description="Timecode (HH:MM:SS)")
    time_in_sec: float = Field(description="Time in seconds")
    time_out_sec: float = Field(description="Time out seconds")
    duration_sec: float = Field(description="Segment duration")

    # Metadata
    rating: int | None = Field(None, ge=1, le=5, description="Hand rating 1-5 stars")
    winner: str | None = None
    hands: str | None = Field(None, description="Poker hands (e.g., AA vs KK)")
    tags: list[str] = Field(default_factory=list)

    # UDM conversion status
    udm: UdmInfo = Field(default_factory=UdmInfo)


class ValidationWarning(BaseModel):
    """Validation warning for a segment"""
    segment_row: int
    type: str
    message: str
    suggested_fix: dict | None = None


class MatchingItem(BaseModel):
    """
    Matching matrix item (Asset Level).
    Represents 1:N relationship between NAS file and segments.
    """
    file_name: str

    # NAS file info (Asset Level - 1 side)
    nas: NasFileInfo | None = None

    # 1:N relationship: segments count
    segment_count: int = Field(0, description="Total sheet records (segments)")
    udm_count: int = Field(0, description="Converted UDM segments")

    # Segments detail (loaded on expand)
    segments: list[SegmentRecord] = Field(default_factory=list)

    # Status
    status: Literal["complete", "partial", "pending", "warning", "no_metadata", "orphan"]
    status_detail: str = Field(description="Status detail (e.g., '15/15', '6/8')")
    warnings: list[ValidationWarning] = Field(default_factory=list)

    # UI state
    is_expanded: bool = False


class MatchingMatrixResponse(BaseModel):
    """Response for GET /api/matching/matrix"""
    total_files: int
    total_segments: int
    matched_files: int
    matched_segments: int
    orphan_records: int
    unmatched_nas: int
    items: list[MatchingItem]


class SourceStats(BaseModel):
    """Statistics for a data source"""
    total_files: int | None = None
    total_records: int | None = None
    unique_files: int | None = None
    total_size_gb: float | None = None
    scanned_at: datetime | None = None
    synced_at: datetime | None = None


class MatchingStatsFiles(BaseModel):
    """File-level matching statistics"""
    complete: int = Field(description="All segments converted")
    partial: int = Field(description="Some segments converted")
    warning: int = Field(description="Has warnings")
    unmatched: int = Field(description="No metadata")
    total_with_metadata: int


class MatchingStatsSegments(BaseModel):
    """Segment-level matching statistics"""
    complete: int
    pending: int
    warning: int
    total: int


class MatchingCoverage(BaseModel):
    """Coverage metrics"""
    archive_to_nas: float = Field(description="Archive Sheet → NAS match rate")
    iconik_to_nas: float = Field(description="Iconik Sheet → NAS match rate")
    nas_to_any_sheet: float = Field(description="NAS → Any Sheet match rate")
    segment_conversion_rate: float = Field(description="Segment UDM conversion rate")


class MatchingSummary(BaseModel):
    """Summary statistics"""
    avg_segments_per_file: float
    max_segments_per_file: int
    min_segments_per_file: int


class MatchingStats(BaseModel):
    """Response for GET /api/matching/stats"""
    sources: dict[str, SourceStats] = Field(
        description="Stats by source (nas, archive_metadata, iconik_metadata)"
    )
    matching: dict[str, MatchingStatsFiles | MatchingStatsSegments | int] = Field(
        description="Matching statistics (files, segments, orphan_records)"
    )
    coverage: MatchingCoverage
    summary: MatchingSummary


class TimeRange(BaseModel):
    """Time range information for a segment"""
    in_tc: str = Field(description="Timecode in")
    out_tc: str = Field(description="Timecode out")
    in_sec: float = Field(description="Time in seconds")
    out_sec: float = Field(description="Time out seconds")
    duration_sec: float


class SegmentMetadata(BaseModel):
    """Metadata for a segment"""
    rating: int | None = None
    winner: str | None = None
    hands: str | None = None
    tags: list[str] = Field(default_factory=list)


class SegmentDetail(BaseModel):
    """Detailed segment information"""
    index: int
    source: Literal["archive_metadata", "iconik_metadata"]
    row_number: int
    time_range: TimeRange
    metadata: SegmentMetadata
    udm: UdmInfo


class FileSegmentsResponse(BaseModel):
    """Response for GET /api/matching/file/{file_name}/segments"""
    file_name: str
    nas: NasFileInfo | None = None
    total_segments: int
    converted_segments: int
    segments: list[SegmentDetail]

File: dashboard/frontend/src/types/matching.ts
// Backend API 응답에 맞춘 타입 정의

// UDM 정보
export interface UdmInfo {
  uuid: string | null
  status: 'complete' | 'pending' | 'warning' | 'error'
}

// 세그먼트 레코드 (1:N 관계의 N)
export interface SegmentRecord {
  row_number: number
  source: 'archive_metadata' | 'iconik_metadata'
  time_in: string
  time_out: string
  time_in_sec: number
  time_out_sec: number
  duration_sec: number
  rating: number | null
  winner: string | null
  hands: string | null
  tags: string[]
  udm: UdmInfo
}

// NAS 파일 정보
export interface NasFileInfo {
  exists: boolean
  path: string
  size_mb: number
  duration_sec: number
  modified_at: string
  inferred_brand: string | null
}

// 검증 경고
export interface ValidationWarning {
  segment_row: number
  type: string
  message: string
}

// 매칭 아이템 (파일 레벨)
export interface MatchingItem {
  file_name: string
  nas: NasFileInfo | null
  segment_count: number
  udm_count: number
  segments: SegmentRecord[]
  status: 'complete' | 'partial' | 'warning' | 'pending' | 'no_metadata' | 'orphan'
  status_detail: string
  warnings: ValidationWarning[]
  is_expanded: boolean
}

// 소스별 통계
export interface SourceStats {
  total_files?: number
  total_records?: number
  unique_files?: number
  total_size_gb?: number
  scanned_at?: string
  synced_at?: string
}

// 매칭 통계
export interface MatchingStats {
  sources: {
    nas: SourceStats
    archive_metadata: SourceStats
    iconik_metadata: SourceStats
  }
  matching: {
    files: {
      complete: number
      partial: number
      warning: number
      unmatched: number
      total_with_metadata: number
    }
    segments: {
      complete: number
      pending: number
      warning: number
      total: number
    }
    orphan_records: number
  }
  coverage: {
    archive_to_nas: number
    iconik_to_nas: number
    nas_to_any_sheet: number
    segment_conversion_rate: number
  }
  summary: {
    avg_segments_per_file: number
    max_segments_per_file: number
    min_segments_per_file: number
  }
}

// 필터 옵션
export interface FilterOptions {
  status?: 'all' | 'complete' | 'partial' | 'warning' | 'pending' | 'no_metadata'
  searchQuery?: string
  sortBy?: 'file_name' | 'segment_count' | 'status'
  sortOrder?: 'asc' | 'desc'
}
